platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.platform.solr.security;
+
+import static javax.ws.rs.core.Response.Status.Family.SUCCESSFUL;
+
+import com.google.common.annotations.VisibleForTesting;
+import ddf.security.encryption.EncryptionService;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import javax.annotation.Nullable;
+import javax.ws.rs.core.Response.StatusType;
+import org.apache.commons.lang.StringUtils;
+import org.apache.felix.utils.properties.Properties;
+import org.codice.ddf.cxf.client.ClientFactoryFactory;
+import org.codice.ddf.cxf.client.SecureCxfClientFactory;
+import org.codice.ddf.platform.util.uuidgenerator.UuidGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolrPasswordUpdateImpl implements SolrPasswordUpdate {
+
+  private static final String ENC_TEMPLATE = "ENC(%s)";
+
+  private static final String SYSTEM_PROPERTIES_FILE = "custom.system.properties";
+  private static final String KARAF_ETC = "karaf.etc";
+  private static final String SET_USER_JSON_TEMPLATE = "{ \"set-user\": {\"%s\" : \"%s\"}}";
+  private static final Logger LOGGER = LoggerFactory.getLogger(SolrPasswordUpdateImpl.class);
+
+  @SuppressWarnings("squid:S2068" /* This constant does hold an actual password */)
+  private static final String SOLR_PASSWORD_PROPERTY_NAME = "solr.password";
+
+  private final ClientFactoryFactory restClientFactoryFactory;
+  private final UuidGenerator uuidGenerator;
+  private java.util.Properties properties;
+  private SolrAuthResource solrAuthResource;
+  private String newPasswordPlainText;
+  private String newPasswordEncrypted;
+  private EncryptionService encryptionService;
+  private StatusType solrResponse;
+  private String newPasswordWrappedEncrypted;
+
+  public SolrPasswordUpdateImpl(
+      UuidGenerator uuidGenerator,
+      ClientFactoryFactory clientFactoryFactory,
+      EncryptionService encryptionService) {
+    this.encryptionService = encryptionService;
+    this.uuidGenerator = uuidGenerator;
+    this.restClientFactoryFactory = clientFactoryFactory;
+    properties =
+        AccessController.doPrivileged(
```

Comment:
❗️ We shouldn't be accessing all properties like that. First it requires permissions for all which is not a good thing as it exposes everything. We should only access those we really need.

In addition, this assumes that the system hash table persist in the system when in fact, it could be changed from under you by someone else via `System.setProperties()`

Reply:
I don't think it is a good idea. If it was just for tracing or reading ... maybe (still not sure it is a good idea since we need to give all permissions for properties) but you are updating them that way and that could be problematic too.

Reply:
I could make a copy of the properties object. That will keep thing from being changed out from under me. However, it means that when the file is written, it will stomp on any changes make by any other threads. Given that no other threads are likely to be touching the solr.password property, I think this is the lesser of two evils.

I don't see a problem with this class having permission for all properties. In fact, I think reducing permissions to individual properties is too granular. The properties object is encapsulated as a private final variable by the class and I'm OK with that.

Reply:
The problem is that to gain access to the whole map, you need to be given read and write to all properties. In security, you should only give out what is required and no more.
My comment was not what if they change your specific properties but the fact that by design, one can replace the whole map with a new one. If that is ever done, you trying to update yours will have no effect.

Reply:
This bundle already has read and read access to the custom.system.properties file. That is already much broader than a single property.

Reply:
I'll give you that one ... however, updates should still not be done through that map. And if that map is changed by something else in the system later, you will be reading and updating a defunct map

Reply:
Direct access to Properties is removed.